woocommerce/woocommerce
# Checkout: Add password strength meter to new field

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:

Adds a password strength meter to the password field on checkout:

![Screenshot 2024-07-04 at 17 31 49](https://github.com/woocommerce/woocommerce/assets/90977/91762efe-5f43-42b4-b200-193a120c761a)

With validation error:

![Screenshot 2024-07-05 at 11 32 52](https://github.com/woocommerce/woocommerce/assets/90977/c7a0a246-8c65-439f-a047-1712a2daee8b)

Key details about the implementation:

- Introduced the ability to pass in the feedback component to `ValidatedTextField` so we can inject the strength meter
- Enqueues `zxcvbn-async` if registration is enabled and the user is logged out. If `zxcvbn` is not present, it will instead use the `check-password-strength` library which is less accurate. We could instead just hide the meter if zxcvbn is not present (feedback?)
- Uses the semantic HTML `meter` component. Browser compatibility was awkward but I tested across Safari, Arc, Firefox. [Firefox does not show the "score" because `:after` is not supported.](https://css-tricks.com/html5-meter-element/)
- I've set a minimum allowed strength of 2 (medium) which will block checkout if not met. Is this sensible default or do we want it lower or unrestricted?

Closes #48628



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. In store Account and Privacy settings, enable registration during checkout and turn off the option to automatically generate passwords
2. As a guest customer, add something to cart and go to checkout.
3. Check the "create account" checkbox
4. Enter a password. Test a variety of passwords and ensure the strength indicator changes 
5. Try to place order with weak password. You should see a validation error.
6. Place order with strong password. Ensure account was created.



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
